rust/scx_utils/examples/mangolog.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+
+// This software may be used and distributed according to the terms of the
+// GNU General Public License version 2.
+use scx_utils::mangoapp::mangoapp_msg_v1;
+
+use anyhow::bail;
+use anyhow::Result;
+use libc::{msgget, msgrcv, IPC_NOWAIT};
+
+use std::mem;
+use std::path::Path;
+use std::thread;
+use std::time::Duration;
+use std::{ffi::CString, io};
+
+fn main() -> Result<()> {
+    // NOTE: the mangoapp file has to be present when gamescope launches in order to read from the
+    // message queue. For example when launching Counter Strike 2 the mangoapp file needs to be in
+    // $HOME/.local/share/Steam/steamapps/common/Counter-String Global Offensive/mangoapp and the
+    // program needs to run in the same directory as well.
+    let file_path = Path::new("mangoapp");
+    if !file_path.exists() {
+        bail!("mangoapp file does not exists");
+    }
+
+    // Create the key for msgget reads,
+    let path = CString::new("mangoapp").unwrap();
+    let key = unsafe { libc::ftok(path.as_ptr(), 65) };
+    if key == -1 {
+        bail!("failed to ftok: {}", io::Error::last_os_error());
+    }
+
+    let msgid = unsafe { msgget(key.try_into().unwrap(), 0) };
+    if msgid == -1 {
+        bail!("msgget failed: {}", std::io::Error::last_os_error());
+    }
+
+    let mut raw_msg: mangoapp_msg_v1 = unsafe { mem::zeroed() };
+
+    loop {
+        let msg_size = unsafe {
+            msgrcv(
+                msgid,
+                &mut raw_msg as *mut _ as *mut libc::c_void,
+                mem::size_of::<mangoapp_msg_v1>() - mem::size_of::<i64>(), // Exclude msg_type
+                0,
+                IPC_NOWAIT,
```

Comment:
I think you want MSG_COPY or'ed with IPC_NOWAIT kinda like here btw: 

https://github.com/likewhatevs/scx_tsatw/blob/c84f0b0b8fcdc61f1e484d3d97ef2bfcb7ef0ea0/util.c#L56

Without that this could mess w/ mangohud output/logs by pulling msgs off of the queue instead of just snooping on them.

Also a wrapper like https://github.com/likewhatevs/scx_tsatw/blob/c84f0b0b8fcdc61f1e484d3d97ef2bfcb7ef0ea0/util.c#L59 might make sense to help with polling (it's kinda wild how fast those IPC msgs are pulled from the queue by mangohud tbh).

Reply:
* Ahh IIRC getting passing MSG_COPY to msgrcv working depends on `-D_GNU_SOURCE` and IDK how one enables that in rust (i.e. maybe a comment noting that this pulls msgs off the queue instead of snooping is sufficient etc.).

Reply:
I think that's right, but for some reason it fails to parse messages with it set:
```
$  ~/git/scx/target/release/examples/mangolog
Error: mangoapp: msgrcv returned -1 with error Function not implemented (os error 38)
```
I'll leave a note in, since this is just example code.